contrib/src/main/scala/akka/stream/contrib/Valve.scala

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.contrib
+
+import akka.stream.stage.{GraphStageLogic, GraphStageWithMaterializedValue, InHandler, OutHandler}
+import akka.stream._
+
+import scala.concurrent.{Future, Promise}
+import scala.util.{Failure, Success, Try}
+
+/**
+  * Pause/ Resume a Flow
+  */
+trait ValveSwitch {
+  /**
+    * Change the state of the valve
+    *
+    * @param mode expected mode to switch on
+    *
+    * @return
+    *         Future[true] if the mode change
+    *         Future[false] if the mode didn't change
+    */
+  def flip(mode: SwitchMode): Future[Boolean]
+}
+
+/**
+  * Valve permit to control a flow (pause/resume) by using a ValveSwitch.
+
+  * After calling [[ValveSwitch#flip(mode: SwitchMode)]] the running instance of the [[Graph]] of [[FlowShape]] that materialized to the
+  * [[Valve]] will change the state of the valve
+  *
+  * After calling [[ValveSwitch#flip(Close)]]: the valve state is changed to Close if a push is requested the current element is saved in a buffer and if a pull is trigger the called will be saved.
+  *
+  * After calling [[ValveSwitch#flip(Open)]]: the valve state is changed to Open if a buffered Element is detected it will be automatically be push and a pull will be trigger if one was requested
+  *
+  * @param mode : state of the valve at the startup of the flow (by default Open)
+  */
+class Valve[A](mode: SwitchMode = SwitchMode.Open) extends GraphStageWithMaterializedValue[FlowShape[A, A], ValveSwitch] {
+
+  override val shape = FlowShape(Inlet[A]("valve.in"), Outlet[A]("valve.out"))
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, ValveSwitch) = {
+    val logic = new ValveGraphStageLogic(shape, mode)
+    (logic, logic.switch)
+  }
+
+  private class ValveGraphStageLogic(shape: Shape, var mode: SwitchMode) extends GraphStageLogic(shape) with InHandler with OutHandler {
+    import shape._
+
+    /**
+      * Buffer the element to push() if the valve was close if a push was requested
+      */
+    var bufferedElement = Option.empty[A]
+
+    val switch = new ValveSwitch {
+
+      override def flip(_mode: SwitchMode): Future[Boolean] = {
+        val promise = Promise[Boolean]()
+
+        val callback = getAsyncCallback[Option[A]]{ element =>
+          val currentMode = mode
+            if (isClose) {
+              element.foreach(push(out, _))
+              bufferedElement = Option.empty
+
+              if (isAvailable(out)){ // if a pull in has been requested we trigger pull(In)
+                pull(in)
+              }
+
+              mode = SwitchMode.Open
+            }
+            else {
+              mode = SwitchMode.Close
+            }
+           promise.success(currentMode != _mode)
```

Comment:
I don't think it does any damage though, because it's not happening concurrently, so buffer will be drained/in not available after the first change to the same state.